drivers/random/random_pool.c

```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright (c) 2017 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <atomic.h>
+#include <kernel.h>
+#include <random.h>
+#include <string.h>
+#include <tinycrypt/aes.h>
+
+#define STIR_THREAD_STACK_SIZE 768
+
+struct random_pool {
+	struct k_sem sem;
+	u32_t pot[4];
+
+#if defined(CONFIG_RANDOM_POOL_SOURCE_CYCLE_COUNTER)
+	atomic_val_t rand32_counter;
+#define RAND32_INC 1000000013
+#endif
+
+#if defined(CONFIG_RANDOM_SOURCE_HARDWARE)
+	struct device *hardware_rng;
+#endif
+
+	struct k_thread stir_thread;
+	u8_t stir_thread_stack[STIR_THREAD_STACK_SIZE];
+};
+
+static inline u32_t rol32(u8_t ch, u32_t n)
+{
+	n &= 31;
+	return ch << n | ch >> (32 - n);
+}
+
+static void stir_bytes(struct random_pool *ctx, void *ptr, size_t size)
+{
+	/* Assumes pot_sem is locked. */
+	static u32_t rotate_count;
+	u8_t *p = ptr;
+	size_t i;
+
+	for (i = 0; i < size; i++) {
+		u32_t v = rol32(*p++, rotate_count++);
+
+		v ^= rol32(ctx->pot[0], rotate_count);
+		v ^= rol32(ctx->pot[1], rotate_count);
+		v ^= rol32(ctx->pot[2], rotate_count);
+		v ^= rol32(ctx->pot[3], rotate_count);
+
+		ctx->pot[rotate_count & 3] = v;
+	}
+}
+
+#if defined(CONFIG_RANDOM_POOL_SOURCE_DEVICES)
+static void init_devices_source(struct random_pool *ctx)
+{
+	struct device *device_list;
+	int devices;
+	int i;
+
+	/* While not random, this ensures that there will be different bits
+	 * for each different device configuration.
+	 */
+
+	device_list_get(&device_list, &devices);
+	for (i = 0; i < devices; i++) {
+		stir_bytes(ctx, device_list[i], sizeof(struct device));
+	}
+}
+#else
+#define init_devices_source(...) do { } while (0)
+#endif
+
+#if defined(CONFIG_RANDOM_SOURCE_HARDWARE)
+static void init_hardware_source(struct random_pool *ctx)
+{
+	u32_t v;
+
+	ctx->hardware_rng = device_get_binding(CONFIG_RANDOM_HARDWARE_NAME);
+	if (!ctx->hardware_rng) {
+		k_panic();
+	}
+
+	random_get_entropy(ctx->hardware_rng, (u8_t *)&v, sizeof(v));
+	stir_bytes(ctx, &v, sizeof(v));
+}
+#else
+#define init_hardware_source(...) do { } while (0)
+#endif
+
+static void pot_stirrer(void *arg1, void *arg2, void *arg3)
+{
+	struct device *device = arg1;
+	struct random_pool *ctx = device->driver_data;
+
+	ARG_UNUSED(arg2);
+	ARG_UNUSED(arg3);
+
+	while (1) {
+		k_sem_take(&ctx->sem, K_FOREVER);
+
+		/* TODO: Maybe try to keep the mutex locked for the least
+		 * amount possible, stirring the pot with stuff from one
+		 * source each time.
+		 */
+
+#if defined(CONFIG_RANDOM_SOURCE_HARDWARE)
+		random_get_entropy(ctx->hardware_rng, (u8_t *)&v, sizeof(v));
```

Comment:
It's possible to have those settings, but `v` won't be used except for the call to `random_get_entropy()`, so that's "safe".  I've changed this code to call a `stir_hardware_rng()` function that will declare, initialize, stir those bytes, and then zero it out explicitly.  All places that calls `random_get_entropy()` on `hardware_rng` are now calls to this function.